base_layer/wallet/src/output_manager_service/output_manager_service.rs

```diff
@@ -0,0 +1,413 @@
+        let pending_transaction = self
+            .pending_transactions
+            .get_mut(&tx_id)
+            .ok_or(OutputManagerError::PendingTransactionNotFound)?;
+
+        self.unspent_outputs
+            .append(&mut pending_transaction.outputs_to_be_spent);
+
+        Ok(())
+    }
+
+    /// Go through the pending transaction and if any have existed longer than the specified duration, cancel them
+    pub fn timeout_pending_transactions(&mut self, period: Duration) -> Result<(), OutputManagerError> {
+        let mut transactions_to_be_cancelled = Vec::new();
+        for (tx_id, pt) in self.pending_transactions.iter() {
+            if pt.timestamp + period < Utc::now().naive_utc() {
+                transactions_to_be_cancelled.push(tx_id.clone());
+            }
+        }
+
+        for t in transactions_to_be_cancelled {
+            self.cancel_transaction(t.clone())?
+        }
+
+        Ok(())
+    }
+
+    /// Select which outputs to use to send a transaction of the specified amount. Use the specified selection strategy
+    /// to choose the outputs
+    fn select_outputs(
+        &mut self,
+        amount: MicroTari,
+        fee_per_gram: MicroTari,
+        strategy: UTXOSelectionStrategy,
+    ) -> Result<Vec<UnblindedOutput>, OutputManagerError>
+    {
+        let mut outputs = Vec::new();
+        let mut total = MicroTari::from(0);
+        let mut fee_without_change = MicroTari::from(0);
+        let mut fee_with_change = MicroTari::from(0);
+
+        match strategy {
+            UTXOSelectionStrategy::Smallest => {
+                self.unspent_outputs.sort();
+                for o in self.unspent_outputs.iter() {
+                    outputs.push(o.clone());
+                    total += o.value.clone();
+                    // I am assuming that the only output will be the payment output and change if required
+                    fee_without_change = Fee::calculate(fee_per_gram, outputs.len(), 1);
+                    fee_with_change = Fee::calculate(fee_per_gram, outputs.len(), 2);
+
+                    if total == amount + fee_without_change || total >= amount + fee_with_change {
+                        break;
+                    }
+                }
+            },
+        }
+
+        if (total != amount + fee_without_change) && (total < amount + fee_with_change) {
+            return Err(OutputManagerError::NotEnoughFunds);
+        }
+
+        Ok(outputs)
+    }
+
+    pub fn pending_transactions(&self) -> &HashMap<u64, PendingTransactionOutputs> {
+        &self.pending_transactions
+    }
+
+    pub fn spent_outputs(&self) -> &Vec<UnblindedOutput> {
+        &self.spent_outputs
+    }
+
+    pub fn unspent_outputs(&self) -> &Vec<UnblindedOutput> {
+        &self.unspent_outputs
+    }
+
+    /// Utility function to determine if an output exists in the spent, unspent or pending output sets
+    pub fn contains_output(&self, output: &UnblindedOutput) -> bool {
+        self.unspent_outputs
```

Comment:
Have you considered using HashSets for these lists? It might simplify (eliminate?) all this logic